### PR TITLE
feat: `VVMDeployer` to deploy older vyper contract

### PR DESCRIPTION
**Pull request by AlbertoCentonze**
### What I did

Added capability to compile and deploy code from older vyper versions using Vyper Version Manager to fetch the required version of the compiler.

### How I did it
- Added `vvm` as a dependency.
- Added a new type of contract deployer `VVMDeployer` that can deploy code compiled by `vvm`.
- Created `load_vvm`, `loads_vvm`, `load_partial_vvm`, `loads_partial_vvm` and exposed them in the public API.

### How to verify it
Unit tests are available

### Description for the changelog
feat: `VVMDeployer` to deploy legacy vyper contracts

### Cute Animal Picture

image

### CODE REVIEW

File: boa/contracts/vvm/vvm_contract.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+
+class VVMDeployer:
+    def __init__(self, abi, bytecode, filename=None, env: Env = None):
+        self.env = env or Env.get_singleton()
+        self.abi = abi
+
+        # extract constructor types from ABI
+        self.types = next(
+            iter(
```
Comment (charles-cooper) [inline boa/contracts/vvm/vvm_contract.py:14]: unnecessary call to `iter`

Comment (charles-cooper) [reply boa/contracts/vvm/vvm_contract.py:14]: i think this expression could also be split over multiple lines for readability

File: boa/contracts/vvm/vvm_contract.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+        self.bytecode = bytecode
+        self.filename = filename
+        self.factory = ABIContractFactory.from_abi_dict(abi)
+
+    def deploy(self, *args):
+        if len(args) != len(self.types):
+            raise ValueError(f"Expected {len(self.types)} arguments, got {len(args)}")
+
+        encoded_args = abi.encode(self.types, args)
```
Comment (charles-cooper) [inline boa/contracts/vvm/vvm_contract.py:32]: we prefer using the internal abi_encode, as it has a consistent API and we can swap out the backend.

File: boa/interpret.py [hunk 3]
```diff
@@ -225,6 +227,42 @@ def load_partial(filename: str, compiler_args=None):
...
 
+def loads_partial_vvm(source_code: str, version: str):
+    # will install the request version if not already installed
+    vvm.install_vyper(version=version)
+    compiled_src = vvm.compile_source(source_code)
+
+    abi = compiled_src["<stdin>"]["abi"]
+    bytecode = compiled_src["<stdin>"]["bytecode"]
+    # remove the 0x prefix
+    bytecode = bytes.fromhex(bytecode[2:])
```
Comment (charles-cooper) [inline boa/interpret.py:238]: use `bytecode.removeprefix("0x")` as hex implementations are not always consistent

**Review by charles-cooper — COMMENTED** [review]
i think maybe instead of introducing new entry points to boa, we should add the vvm compilation directly into the original `load/loads` machinery.

additionally, we could auto-detect the compiler version so that we know which version to call vvm with. it's a bit of overhead (i just checked, and calling `pre_parse` is about 12.5ms for 850 lines), but worth it so the user doesn't need to double specify the compiler version (both in the pragma and also in boa code). or maybe specifying the compiler version from boa is a good thing?

**Comment by charles-cooper** [comment]
with straight regex it could be much faster --
```python
In [4]: r = re.compile(r"\s*#\s*pragma\s+version\s+\d+\.\d+\.\d+")
In [6]: r.match(s)
Out[6]: <re.Match object; span=(0, 23), match='# pragma version 0.3.10'>
In [7]: %timeit r.match(s)
341 ns ± 24.5 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```

File: boa/interpret.py [hunk 4]
```diff
@@ -225,6 +227,42 @@ def load_partial(filename: str, compiler_args=None):
...
+
+    return VVMDeployer(abi, bytecode, filename="<stdin>")
+
+
+def loads_vvm(source_code: str, version: str, *args):
```
Comment (DanielSchiavini) [inline boa/interpret.py:243]: I agree with Charles that this should be part of the standard API

File: pyproject.toml [hunk 5]
```diff
@@ -24,6 +24,9 @@ dependencies = [
     "pytest",
     "pytest-cov",
 
+    # required to compile older versions of vyper
+    "vvm",
```
Comment (DanielSchiavini) [inline pyproject.toml:28]: maybe this should be part of some optional dependencies? :thinking:

**Comment by DanielSchiavini** [comment]
> with straight regex it could be much faster --
@charles-cooper  Isn't the pragma guaranteed to be in the beginning of the file? As soon as we encounter anything but comments the rest of the file can be ignored.

**Comment by charles-cooper** [comment]
> > with straight regex it could be much faster --
> 
> @charles-cooper Isn't the pragma guaranteed to be in the beginning of the file? As soon as we encounter anything but comments the rest of the file can be ignored.

no it's not, the proper way is to use pre_parse. for example the regex method can produce a different result than the tokenization method. but i think it's ok, it's a fast way to get a heuristic of which compiler to use.

we could do something more failsafe like searching for *all* the regex matches, and in the rare case that the regex returns more than one match we fall back to the tokenization method. but i think it's overkill until somebody actually runs into the issue

File: pyproject.toml (hunk 5 above)
Comment (charles-cooper) [reply pyproject.toml:28]: i don't think we need to make it optional. vvm is under vyper-controlled supply chain, it is a small dependency, and it has no new external dependencies compared to boa (see https://github.com/vyperlang/vvm/blob/f7b4de84e070062eb4519b2fb5ab0ee8e72269e0/setup.py#L20)

File: boa/interpret.py [hunk 6]
```diff
@@ -225,6 +227,42 @@ def load_partial(filename: str, compiler_args=None):
         )
 
 
+def loads_partial_vvm(source_code: str, version: str):
+    # will install the request version if not already installed
+    vvm.install_vyper(version=version)
+    compiled_src = vvm.compile_source(source_code)
```
[the remark this answers is not in the document]
Comment (charles-cooper) [reply boa/interpret.py:233]: mm yes, but i think this is an ok starting point

File: boa/contracts/vvm/vvm_contract.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
+from eth_abi import abi
+
+from boa import Env
+from boa.contracts.abi.abi_contract import ABIContractFactory
+
+
+class VVMDeployer:
+    def __init__(self, abi, bytecode, filename=None, env: Env = None):
+        self.env = env or Env.get_singleton()
```
[the remark this answers is not in the document]
Comment (charles-cooper) [reply boa/contracts/vvm/vvm_contract.py:9]: hmm we can leave it optional. `None` indicates that env selection should be deferred to deploy time.

**Pull request merged**